DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs

```diff
@@ -78,5 +82,52 @@ public IActionResult DeactivateDelegate(int delegateId)
 
             return RedirectToAction("Index", new { delegateId });
         }
+
+        [HttpGet]
+        [Route("{customisationId:int}/Remove")]
+        public IActionResult ConfirmRemoveFromCourse(int delegateId, int customisationId)
+        {
+            var centreId = User.GetCentreId();
+            var delegateUser = userDataService.GetDelegateUserCardById(delegateId);
+            var course = courseDataService.GetCourseNameAndApplication(customisationId);
+            if (delegateUser == null || delegateUser.CentreId != centreId || course == null)
+            {
+                return new NotFoundResult();
+            }

```

Comment:
We should be able to get rid of this by using the new service filters:
VerifyAdminUserCanAccessDelegateUser
VerifyAdminUserCanAccessCourse
The first one already looks to be applied at the controller level. The second can probably be applied to the two new controller actions. 

However, you might want to double check it does what you want it to do. Currently, the manual check in this method only determines that the course exists, while the service filter also checks the admin user has access to edit the course. That might not be applicable here if the admin should be able to remove delegates from AllCentre courses that they don't have permission to edit.

Looking at the HLD comments, Stella mentioned that the Admin Category Id should be checked against to make sure the admin user can remove the user from the course. Seeing as how the admin ability is directly restricted by course properties, I think the course service filter should be fine to apply here (it should also check against the admin category id for you). Thoughts?

Reply:
I think Stella's comments on the HLD say we want to have the check, following her discussion with Kevin. If it's not the desired behaviour they can let us know and we can change it.